packages/malloy/src/lang/test/parse.spec.ts

```diff
@@ -303,11 +303,494 @@ describe('model statements', () => {
+            group_by: c is 1
+            aggregate: c2 is sum(sum(ai))
+          }`).compileToFailWith(
+            'Aggregate expression cannot be aggregate',
+            "Cannot define 'c2', value has unknown type"
+          );
+        });
+        test(
+          'can use field def in group_by, preserved over refinement',
+          modelOK(`query: a1 is a -> {
+            group_by: c is 1
+          }
+          query: -> a1 {
+            order_by: c
+          }`)
+        );
+        test(
+          'can use field ref in group_by, preserved over refinement',
+          modelOK(`query: a1 is a -> {
+            group_by: c is astr
+          }
+          query: -> a1 {
+            order_by: c
+          }`)
+        );
+      });
+    });
+    describe('function typechecking', () => {
+      test(
+        'use function correctly',
+        modelOK(`query: a -> {
+          group_by: s is concat('a', 'b')
+        }`)
+      );
+      test('function no matching overload', () => {
+        expect(`query: a -> {
+          group_by: s is floor('a', 'b')
+        }`).compileToFailWith(
+          'No matching overload for function floor(string, string)',
+          "Cannot define 's', value has unknown type"
+        );
+      });
+      test(
+        'can select different overload',
+        modelOK('query: a -> { group_by: s is concat() }')
+      );
+      test(
+        'can pass different expression types',
+        modelOK(`query: a -> {
+          group_by: f1 is sqrt(1)
+          aggregate: f2 is sqrt(count())
+          aggregate: f3 is sqrt(all(count()))
+          calculate: f4 is sqrt(lag(f1))
+          calculate: f5 is sqrt(lag(count()))
+        }`)
+      );
+      test(
+        'function return type correct',
+        modelOK(`query: a -> {
+          group_by: s is floor(1.2) + 1
+        }`)
+      );
+      test('function return type incorrect', () => {
+        expect(`query: a -> {
+            group_by: s is floor(1.2) + 'a'
+        }`).compileToFailWith(
+          "Non numeric('number,string') value with '+'",
+          "Cannot define 's', value has unknown type"
+        );
+      });
+      test(
+        'can use output value in calculate',
+        modelOK(`query: a -> {
+          group_by: x is 1
+          calculate: s is lag(x)
+        }`)
+      );
+      test('cannot use output value in group_by', () => {
+        expect(`query: a -> {
+          group_by: x is 1
+          group_by: y is x
+        }`).compileToFailWith(
+          "'x' is not defined",
+          "Cannot define 'y', value has unknown type"
+        );
+      });
+      test('lag can check that other args are constant', () => {
+        expect(`query: a -> {
+          group_by: x is 1
+          calculate: s is lag(x, x)
+        }`).compileToFailWith(
+          // TODO improve this error message
+          "Parameter 2 ('offset') of lag must be literal or constant, but received output"
```

Comment:
We might want to improve this error message, because "output" is not an obvious term.